docs/proposals/action_based_roles.md

```diff
@@ -0,0 +1,156 @@
+to invoke a particular piece of a code.
+
+## Rationale
+
+Each component lives through defined life-cycle phases. At the beginning
+a component is installed. As time passes new version of the component is
+available resulting in component upgrade. Under certain conditions one needs
+to carry additional operations to update successfully such as migration,
+backup or post-verification.
+
+Given each of the operations are applied on the same component, some of the operation
+underlying tasks, variables and other Ansible resources are shared. E.g. templates
+of service files, handlers, sequence of installation or configuration tasks.
+
+Additionally, some operations can be unified into one and the proper behavior hidden
+inside a role as implementation details. E.g. component upgrade can be equivalent
+to component installation up to some differences. The role dealing with the installation
+should then decide if the actual installation or rather an update is required.
+
+In practice, it is natural to create a role that deals with an installation as
+the first step. Later, role that deals with an upgrade is implemented. Not rarely
+followed by implementation of other operations. At the end one can end up
+with a set of roles, each responsible for a single operation. Naturally,
+the layout leads to code duplication and various efforts to minimize it.
+
+## Design
+In case a set of roles conforms to the following conditions:
+
+* each role is responsible for a distinct operation (e.g. installation, upgrade, migration)
+* all operations are applied on the same component
+* the operations share common code base and Ansible resources
+
+consolidate all roles into new one, providing a mechanism to select specific
+operation to apply when the new role is invoked. Since Ansible 2.2 one can
+use `include_role` module and perform file-based operation selection:
+
+```yaml
+include_role:
+  name: <role>
+tasks_from: <operation>
+```
+
+Additionally, one can define a minimal set of actions that each action-based
+role should provide:
+* pre-checks: to check if the role has all necessary conditions fulfilled
+* post-checks: to check a role ended successfully
+
+### Layout
+
+Proposed layout of an action-based role:
+
+```sh
+$ ls <role>/tasks
+<action1_dir>
+<action2_dir>
+...
+<action1>.yml
+<action2>.yml
+...
+```
+
+The role `tasks` directory consists of:
+* one directory per each action (action directory)
+* one yml file per each action that includes a file from corresponding action directory
+* additional yml files needed by the role (e.g. task files shared across actions)
+
+Templates and other resources location are not affected by the layout.
+If reasonable, the action directory layout can be duplicated inside the `templates` directory.
+
+**Expected content of an action file**:
+
+```yaml
+# content of tasks/<action>.yml
+
+# include a main file from action directory
+- name: Running <action>
+  include: <action>/main.yml
```

Comment:
I don't prefer to have a main.yml inside the action folder.

I would prefer the following:
```yaml
# tasks/<action-1>.yml
- include common_thing.yml
- some_task:
    some_param: {{ some_var }}
- include <action-1>/do_something.yml
- include <action-1>/do_next.yml
```

The reason being, for users, I don't want people to open a file that is simply a pass-through, and then have to open <action-1>/main.yml, which may or may not include some tasks, which then may include other files, possibly from the parent directory.  It's too much searching.

Putting your series of includes inside of tasks/action-1.yml allows a high-level overview of what that portion of the role (or as I've been calling them, pseudo-role) is doing, and allows importing of shared elements without having to do relative parent includes (eg. include ../common/some_thing.yml)

Reply:
Though it require to re-understand each role, I like this approach. The included file was just an example but I can mention that in the proposal as well.

Reply:
For the most part, I think this will be a drop-in change.  You're essentially renaming old_role/main.yml -> new_role/action_1.yml

Then, you just need to move anything that wasn't the old role's main.yml in the action1 directory, and just refactor the include statements from action_1.yml to include the new folder.  Everything else can be exactly the same.

Then, later, after some common things are identified among the roles that are being consolidated, we can refactor those pieces to be used in a common place somewhere.

Reply:
Expected content of the action file updated